### ISSUE TITLE
Is it a bug?

### ISSUE THREAD

**Issue by PotatoEt** [aaronWool/keygnet#3]
I find a error,but not sure I sould change it. it's in train.py. Wass loss changed as:
```
def cal_loss_wass_gp(input,segments,vote_type,model,device,lambda_,training):
    '''
    input: keypoint tensor [b,n,3]
            b-batch
            n-number of points
            3-x,y,z
    segment: segment point cloud tensor [b,N,3]
            b-batch
            N-number of points
            3-x,y,z    
    vote_type: 0-radii
               1-vector
               2-offset    
    return wass loss
    ''' 
    b = input.size()[0]   
    n = input.size()[1]
    N = segments.size()[2]
    sum = 0
    iter = 0
    for i in range(b):
        kpoints = input[i]
        segment = segments[i]
        segment = segment.permute(1,0)
        #if vote_type==0:
            #votes = torch.zeros(n,N)
            #for j in range(n):
                #kpoint=kpoints[j]
                #votes[j]=((kpoint[0]-segment[:,0])**2+(kpoint[1]-segment[:,1])**2+(kpoint[2]-segment[:,2])**2)**0.5
        #else:
        #if vote_type!=0:
        votes = torch.zeros(n,N,3).cuda(device) 
        norms = torch.zeros(n,N).cuda(device)
        sorted_norms = torch.zeros(n,N).cuda(device)
        sorted_indices = torch.zeros(n,N,dtype=torch.long).cuda(device)
        for j in range(n):
            kpoint=kpoints[j]
            #print(kpoint.shape)
            #print(segment.shape)
            un_squeezed_kpt = torch.unsqueeze(kpoint,0)
            un_squeezed_kpt = un_squeezed_kpt.repeat(N,1)
            #print(un_squeezed_kpt.shape)
            #print(segment[:,:3].shape)
            xyz=segment[:,:3]
            offsets=torch.sub(un_squeezed_kpt,xyz)
            norm=torch.norm(offsets,dim=1)
            #print(norm.shape)
            #print(norms.shape)
            norms[j]=norm
            #norms[j]=((kpoint[0]-segment[:,0])**2+(kpoint[1]-segment[:,1])**2+(kpoint[2]-segment[:,2])**2)**0.5
            sorted_norms[j],sorted_indices[j]=torch.sort(norms[j])
            if vote_type!=0:
                votes[j] = offsets
            if vote_type==1:
                #norm = ((kpoint[0]-segment[:,0])**2+(kpoint[1]-segment[:,1])**2+(kpoint[2]-segment[:,2])**2)**0.5
                repeated_norm = norm.repeat(3,1)
                repeated_norm = repeated_norm.permute(1,0)
                vectors = torch.div(offsets,repeated_norm)
                votes[j] = vectors
            if vote_type!=0:
                #cross product
                vote = votes[j]
                cross = torch.cross(vote,un_squeezed_kpt)
                cross_copy = cross.detach().clone()
                norms[j] = torch.norm(cross_copy,dim=1)
                sorted_norms[j],sorted_indices[j]=torch.sort(norms[j])
                #votes[j] = votes[j,sorted_indices[j],:]

        for j in range(0,n-1):
            for k in range(j+1,n): 
                # if vote_type==0:
                wass_dist=torch.mean(torch.abs(sorted_norms[j]-sorted_norms[k]))
                sum+=wass_dist
                # else:
                #     wass_dist=torch.mean(torch.abs(votes[j]-votes[k]))
                #     sum+=wass_dist
                iter+=1
    l_wass=sum/iter
    if training:
        idx=np.random.choice(np.arange(b),b,replace=False)
        shuffled = segments[idx,:,:]
        #print(shuffled.size())
        #gradient penalty
        shuffled = Variable(shuffled, requires_grad=True)
        prob_shuffled = model(shuffled)
        #print(prob_shuffled.get_device())
        # calculate gradients of probabilities with respect to examples
        gradients = autograd.grad(outputs=prob_shuffled, inputs=shuffled,
                               grad_outputs=torch.ones(
                                   prob_shuffled.size()).cuda(device) if args.cuda else torch.ones(
                                   prob_shuffled.size()),
                               create_graph=True, retain_graph=True)[0]
        `#print(gradients.get_device())`
        grad_penalty = ((gradients.norm(2, dim=1) - 1) ** 2).mean() * lambda_
        #print('wass: ', l_wass)
        #print('gp: ', grad_penalty)
        l_wass+=grad_penalty
    return l_wass 
```
I just change the wass loop
```
        for j in range(0,n-1):
            for k in range(i+1,n): 
                # if vote_type==0:
                wass_dist=torch.mean(torch.abs(sorted_norms[i]-sorted_norms[k]))
                sum+=wass_dist
                # else:
                #     wass_dist=torch.mean(torch.abs(votes[i]-votes[k]))
                #     sum+=wass_dist
                iter+=1
```
to
```
        for j in range(0,n-1):
            for k in range(j+1,n): 
                # if vote_type==0:
                wass_dist=torch.mean(torch.abs(sorted_norms[j]-sorted_norms[k]))
                sum+=wass_dist
                # else:
                #     wass_dist=torch.mean(torch.abs(votes[j]-votes[k]))
                #     sum+=wass_dist
                iter+=1
```
